### ISSUE TITLE
aws-s3-deployment: deployment fails if asset folder doesn't exist yet

### ISSUE THREAD

**Issue by dbartholomae** [aws/aws-cdk#20185]
### Describe the bug

When using `aws-s3-deployment/Source` to bundle an asset, if the folder of the asset doesn't exist yet, it fails. Given that the asset is only bundled _after_ that check, the check is failing even though the asset will be created correctly later.

### Expected Behavior

It should bundle without an error.

### Current Behavior

```
      throw new Error(`Cannot find asset at ${this.sourcePath}`);
            ^
Error: Cannot find asset at D:\repos\source\admin\sfmc-custom-activity\build
    at new AssetStaging (D:\repos\source\common\temp\node_modules\.pnpm\aws-cdk-lib@2.20.0_constructs@10.0.121\node_modules\aws-cdk-lib\core\lib\asset-staging.ts:165:13)
    at new Asset (D:\repos\source\common\temp\node_modules\.pnpm\aws-cdk-lib@2.20.0_constructs@10.0.121\node_modules\aws-cdk-lib\aws-s3-assets\lib\asset.ts:131:21)
    at Object.bind (D:\repos\source\common\temp\node_modules\.pnpm\aws-cdk-lib@2.20.0_constructs@10.0.121\node_modules\aws-cdk-lib\aws-s3-deployment\lib\source.ts:108:23)
    at D:\repos\source\common\temp\node_modules\.pnpm\aws-cdk-lib@2.20.0_constructs@10.0.121\node_modules\aws-cdk-lib\aws-s3-deployment\lib\bucket-deployment.ts:316:83
    at Array.map (<anonymous>)
    at new BucketDeployment (D:\repos\source\common\temp\node_modules\.pnpm\aws-cdk-lib@2.20.0_constructs@10.0.121\node_modules\aws-cdk-lib\aws-s3-deployment\lib\bucket-deployment.ts:316:51)
    at new SfmcCustomActivityStack (D:\repos\source\admin\sfmc-custom-activity\infrastructure\SfmcCustomActivityStack.ts:77:5)
    at new SfmcCustomActivityPipelineStack (D:\repos\source\admin\sfmc-custom-activity\infrastructure\SfmcCustomActivityPipelineStack.ts:72:5)
    at Object.<anonymous> (D:\repos\source\admin\sfmc-custom-activity\bin\production.ts:16:1)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
```

### Reproduction Steps

```
    const bundle = Source.asset("./build");
```
while a `build` folder doesn't exist yet.

### Possible Solution

If possible, just remove the check from `AssetStaging`, as it makes the wrong assumption that the asset should already exist when its constructor is called.

### CDK CLI Version

2.20.0

### Node.js Version

16.14.2

### OS

Windows 11

### Language

Typescript

**Comment by peterwoodworth** [comment]
I don't think this is a bug, or something we can implement - this is how it happens because we generate the asset hash at synth time based on the contents of the file https://github.com/aws/aws-cdk/blob/v1.157.0/packages/@aws-cdk/core/lib/asset-staging.ts#L197-L204

**Comment by dbartholomae** [comment]
@peterwoodworth I do believe it is. There are three different hash types: One hashes the source, one hashes the output, and one allows users to manually provide a hash. In all cases except for the first, this fails.
In my case, I use an output hash. Here's an excerpt from the code:

```
const bundle = Source.asset("./build", {
      // Needed to make sure we rebuild when env variables change
      assetHashType: AssetHashType.OUTPUT,
      bundling: {
        command: [
          "sh",
          "-c",
          'echo "Docker build not supported. Please install esbuild."',
        ],
        image: DockerImage.fromRegistry("alpine"),
        local: {
          tryBundle(outputDir: string) {
            try {
              execSync("esbuild --version");
            } catch {
              console.log("Esbuild not installed");
              return false;
            }
            execSync("npm run build-react-app", {
              env: {
                ...process.env,
              },
            });
            fs.copySync("./build", outputDir);
            return true;
          },
        },
      },
    });

    new BucketDeployment(this, "DeployWebsite", {
      sources: [bundle],
      destinationBucket: bucket,
      destinationKeyPrefix,
      retainOnDelete: false,
    });
```

The `build` folder does not exist before `tryBundle` is called, but it does after, which is also the point in time when the hash is calculated. But the check in `AssetStaging` happens before, which causes the problem.

**Comment by mrgrain** [comment]
@dbartholomae In this code example, it looks like your source is actually `.` (the root) and `./build` is a temporary directory that would not be needed if you were to update the `esbuild` and `npm run build-react-app` commands to put any outputs directly into `outputDir`. Also, I think you could change `./build` to any other value here and it would still behave exactly the same.

Is my understanding correct?

--- 

Admittedly this isn't very obvious on `Source.asset()` but `AssetProps.path` makes its expectations quite clear:
> The disk location of the asset.
The path should refer to one of the following:
- A regular file or a .zip file, in which case the file will be uploaded as-is to S3.
- A directory, in which case it will be archived into a .zip file and uploaded to S3.

[Source](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_assets.AssetProps.html#path)


--- 
The existing Asset contract is overloaded. It provides staging by a copy mechanism and by bundling. To be honest, I don't like how it's currently done and it needs an overhaul.

Specifically both mechanism are detached from each other and the props they are using (see note on [exclude](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3_assets.AssetProps.html#exclude)).

For path it's similar. The assumptions that it's an existing directory is required for the copy mechanism. You're right though, it's not currently used for bundling. We might want to make use of the path and its contents in the future, e.g. for watch mode. That shouldn't stop us from addressing a bug though.

--- 
tl;dr I think the intention for `path` was always to point at the source data, not a compiled output. This isn't currently conveyed explicitly enough though. While we are not in all situations making use of `path` at the moment, I'm not convinced yet the provided examples make a strong case to give up that assumptions.

**Comment by dbartholomae** [comment]
> @dbartholomae In this code example, it looks like your source is actually `.` (the root) and `./build` is a temporary directory that would not be needed if you were to update the `esbuild` and `npm run build-react-app` commands to put any outputs directly into `outputDir`.

Yes, that's correct. The reason for the temp folder is to make sure that I can create a reproducible hash which only relies on files relevant to the deployment (the root also contains other files, potentially dead code that is optimized away by esbuild, as well as a lot of symlinked files).

> Also, I think you could change `./build` to any other value here and it would still behave exactly the same.
> Is my understanding correct?

As long as I both change the location where the asset is copied to as well as the location that is used to calculate the hash, yes.

> tl;dr I think the intention for `path` was always to point at the source data, not a compiled output. This isn't currently conveyed explicitly enough though. While we are not in all situations making use of `path` at the moment, I'm not convinced yet the provided examples make a strong case to give up that assumptions.

If that really was the intention, then I would challenge it as it is at-odds with some base ideas behind AWS CDK. Pointing to "the source data" only works if all of the source data is neatly organized in one folder, with nothing else in it. But the AWS CDK follows the idea the application code should be co-located with the infrastructure code.
A good example is `NodejsFunction`, which by default assumes that the lambda handler code is in the same directory as the construct it is used in. The current implementation has the consequence that if I have e.g. a lambda function and another resource defined in the same package, the lambda hash will change if I change the other resource, because all files in the project are included in the hash.

I also don't really see any benefit in testing whether a folder exists at that point. The same error can still be shown after the local bundling was executed without any negative consequences on future use cases like watch mode.

**Comment by mrgrain** [comment]
@dbartholomae I'm confused. In the provided example you do not calculate a hash. You have set it to `AssetHashType.OUTPUT` which specifically uses the contents of  `outputDir` to create the hash.

---

I think `NodejsFunction` behaves exactly as I described the intention: You point it at the (uncompiled) source JS/TS file. Either by naming convention, or by setting an explicit entry file. Its bundling feature than takes this source, does some work and puts the result into `outputDir`. The  `outputDir` and its contents are then used to calculate the hash.

--- 

> The current implementation has the consequence that if I have e.g. a lambda function and another resource defined in the same package, the lambda hash will change if I change the other resource, because all files in the project are included in the hash.

Which implementation has this problem? Files from the project are only included in the hash if 

From my experience maintaining @mrgrain/cdk-esbuild I have never encountered that and that package defaults to use the root as `path` (see https://github.com/mrgrain/cdk-esbuild/blob/main/src/asset.ts#L105)

--- 

Does this issue mainly come from [Create React App](https://create-react-app.dev/) putting the compiled output into `./build` by default?

**Comment by dbartholomae** [comment]
After diving a bit deeper through the AWS code, I think I come to the conclusion that you are right and that basically in the first line `const bundle = Source.asset("./build", {` I can replace `./build` with any directory, as it is not used for bundling at all.
What confused me is that `./build` is used in multiple places, serving different functions, and the first one is basically not used at all when bundling. Not sure why it took me multiple replies to understand that, your explanation @mrgrain was quite clear 😅
I'll think a bit whether there is a way to make this clearer in the docs, but will close the issue for now.
